spinalcordtoolbox/vertebrae/core.py

```diff
@@ -194,28 +195,32 @@ def vertebral_detection(fname, fname_seg, contrast, param, init_disc, verbose=1,
         # assign new current_z and disc value
         if direction == 'superior':
             try:
-                approx_distance_to_next_disc = list_distance[list_disc_value_template.index(current_disc - 1)]
+                approx_distance_to_next_disc = list_distance[list_disc_value_template.index(current_value - 1)]
             except ValueError:
                 logger.warning('Disc value not included in template. Using previously-calculated distance: %s', approx_distance_to_next_disc)
             # assign new current_z and disc value
             current_z = current_z + approx_distance_to_next_disc
-            current_disc = current_disc - 1
+            current_value = current_value - 1
         elif direction == 'inferior':
             try:
-                approx_distance_to_next_disc = list_distance[list_disc_value_template.index(current_disc)]
+                approx_distance_to_next_disc = list_distance[list_disc_value_template.index(current_value)]
             except ValueError:
                 logger.warning('Disc value not included in template. Using previously-calculated distance: %s', approx_distance_to_next_disc)
             # assign new current_z and disc value
             current_z = current_z - approx_distance_to_next_disc
-            current_disc = current_disc + 1
+            current_value = current_value + 1
```

Comment:
Ah! So, there was a small, curious thing that caught my attention:

* The error I observed from [`test_sct_label_vertebrae_initial_disc_too_low`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3775#pullrequestreview-950071054) prior to your try/except changes was `ValueError: min() arg is an empty sequence`
* However, the user's error from [the reported traceback](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3775#issue-1210242621) was an `UnboundLocalError`

Either way, your fix _does_ address the issue (by catching a `current_value` that doesn't exist in the template). But, the test itself produced slightly different behavior _prior_ to your changes. Curious!

I tried investigating further by tracing your test and seeing how it differs from the user's report. And it turns out, when we supply `-initz 40,-1` as we do in that test, the `current_value` actually gets parsed as _254_???

![image](https://user-images.githubusercontent.com/16181459/164730394-03890e15-dd64-4b5b-8d3e-d07188fa7c34.png)

I think this is because of this line:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/scripts/sct_label_vertebrae.py#L383

Which creates a `uint8` label image from the `initz` argument (i.e. the image is bound to `[0, 255]`). So, I'm guessing that by supplying -1, it wraps around and creates a labeled image with the value `254`? (I tried changing the test to -2 and it produced an image with label value `253`.)

But, as I mentioned, your check still catches the value `254` as it would `-1`, so... I'm not sure. Is this something we want to investigate while we're here? If not, this has my approval. :)

Reply:
Oh wait, I can just quickly update `create_labels_along_segmentation` to not use `uint8`, to see what would happen if we did in fact supply `-1`. One sec. :V

Reply:
So, if I change this line to use `dtype=np._int`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/labels.py#L111

We get the `-1` label in the image as expected. However, then when the image gets dilated here:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/scripts/sct_label_vertebrae.py#L402

The dilation causes the `-1` value to disappear. So, again, another little factor preventing your `-1` test from actually doing what it's supposed to. :P

Reply:
Ah, we use [`skimage.morphology.dilation`](https://scikit-image.org/docs/dev/api/skimage.morphology.html#skimage.morphology.dilation) for our dilate function, which inherently returns a `uint8` image.

So... I guess negative label values are impossible via `fname_labelz`? But... the user's error certainly suggests that it's possible...

Reply:
Again though, just to re-emphasize, I think we've covered all of our bases with the new corner cases (0, 19, 20, 21), so fixating on the `-1` test isn't really that important?

I just wanted to highlight that the test wasn't doing what we thought it was doing. :P

Reply:
That's a great investigation! I didn't think about numpy's wrap-around integer types here, but it's a good source of surprises and bugs. And you're right that image dilation does surprising things with negative numbers.

But! There's one more place where negative numbers can creep in :stuck_out_tongue: After straightening of `fname_labelz`, the label is read from here:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/scripts/sct_label_vertebrae.py#L414
and then we subtract 1 before handing to `vertebral_detection`:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/scripts/sct_label_vertebrae.py#L433
Note that by this point `init_disc[1]` is an honest-to-goodness Python int, because it's explicitly cast here:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/bad57807faf05643e2697c911f3403c4f57d9f66/spinalcordtoolbox/vertebrae/core.py#L298

So, I think `-1` is in fact the *only* possible negative value, and it happens with `-initz *,0` (or I guess, any positive or negative multiple of 256?). This also explains why you saw 254 instead of 255 when using `-initz *,-1`.

----

~~Now, I want to argue that the `min() arg is an empty sequence` is actually already addressed by this PR.~~ The error is raised on this line after breaking out of the main loop:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/3f7832ad5670fc6cb875c8b413a37c15a1e5166c/spinalcordtoolbox/vertebrae/core.py#L232
The argument to `min` can only be empty if this code in the loop never executes:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/3f7832ad5670fc6cb875c8b413a37c15a1e5166c/spinalcordtoolbox/vertebrae/core.py#L168-L176
And the only `break` before that is here:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/3f7832ad5670fc6cb875c8b413a37c15a1e5166c/spinalcordtoolbox/vertebrae/core.py#L145-L151
However, ~~the new error checking code prevents this from happening on the first iteration,~~ because we now check whether the initial `current_value` is there:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/3f7832ad5670fc6cb875c8b413a37c15a1e5166c/spinalcordtoolbox/vertebrae/core.py#L134-L136
Edit: the check which would actually prevent this is `current_value in range(-1-len(list_disc_value_template), len(list_disc_value_template))`. But for the PAM50 template, `list_disc_value_template == [0, 1, 2, ..., 19]`, so the new error checking code accidentally catches this condition.

I think we're hitting diminishing returns here; the real solution is not to add more error checking code, but to actually fix the underlying code after the 5.6 release. So, I think I'll add some comments and/or docstrings for the new corner-case tests and call this PR done, if that's ok with you.

Reply:
>  So, I think I'll add some comments and/or docstrings for the new corner-case tests and call this PR done, if that's ok with you.

Perfectly reasonable to me! That was very much my feelings too after writing https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3775#discussion_r856302643.